bootchooser: add systemd-boot bootloader support

obbardc opened the pull request:
Add some really basic support to set the EFI variable for systemd-boot's default boot choice EFI variable.

This is currently set when "rauc status mark-good" is called; meaning that once the system has started successfully the default boot choice EFI variable will be updated to reflect from the next boot onwards.

---

This PR needs (at least):
- [ ] rebasing onto latest main (currently based on v1.2) & testing
- [ ] implementation for `r_boot_get_primary`
- [ ] implementation for `r_boot_set_primary`
- [ ] test cases
- [ ] documentation!

For context, this is from a project which we converted to use `systemd-boot` for its bootloader and is working quite well. I am creating a draft PR to see if there is any interest in accepting this into upstream, I can do all of the work to get it in.

obbardc commented on src/bootchooser.c:
```diff
@@ -1456,6 +1457,51 @@ static gboolean efi_get_state(RaucSlot* slot, gboolean *good, GError **error)
+	}
+
+	/* Set the default boot choice to this slot name */
+	/* TODO: allow the slot name to be configurable */
```
Should add a config option for this.

jluebbe replied:
Shouldn't this be the slot's bootname?

jluebbe commented:
How would this interact with https://systemd.io/AUTOMATIC_BOOT_ASSESSMENT/, which is also implemented by systemd-boot?

jluebbe commented:
This would need to be rebased, as https://github.com/rauc/rauc/pull/1636 has split the bootloader backends into different files.

obbardc committed:
bootchooser: add systemd-boot bootloader support

Add a new bootloader backend for systemd-boot using boot entry
configuration and EFI variables to determine and update the active
boot slot.

Wire up get_state, set_state, get_primary and set_primary to the newly
introduced bootchooser. Add mock stubs and a test in test/bootchooser.c
covering the full mark-good/mark-bad/set-primary flow.

Add documentation to help users integrate systemd-boot and RAUC into
their system.

Signed-off-by: Christopher Obbard <christopher.obbard@linaro.org>

obbardc commented:
@jluebbe I rebased onto master and updated the PR. It is currently untested (apart from unit tests in RAUC); so take with a punch of salt.
